### DISCUSSION TITLE
prefer-import-index

### DISCUSSION THREAD

**Discussion by giogonzo** [buildo/eslint-plugin-fp-ts#13] [Ideas for new rules]
OK:

```ts
import { taskEither } from 'fp-ts';
```

NOK:

```ts
import { taskEither } from 'fp-ts/TaskEither';
```

**Comment by gabro** [comment]
Good one, I wonder how can we discriminate that. We can't just ban imports from submodules altogether, maybe can just list all the instances and be done with it? Or only allow importing types from the submodules?

**Reply by giogonzo** [reply]
> Or only allow importing types from the submodules?

Some submodules are also exporting legit values, e.g. `fp-ts/function` (unless you start to suggest `function.pipe`).
Listing the instances seems safer although it requires maintenance

**Reply by gabro** [reply]
right, however I feel we should also avoid importing utils from modules, like:

```ts
// error
import { some } from 'fp-ts/Option'

const v = some(2)

// ok
import { option } from 'fp-ts'

const v = option.some(2)
```

Listing all the utils does not seem like an appealing option... Maybe we can special-case the module that have legit exports such as `function`?

**Reply by giogonzo** [reply]
Makes sense.
If the scope is only "instances": if there is a way to know that `X` imported from `'fp-ts/X'` is also exported by `'fp-ts'` itself, then we could only suggest the fix based on this info. But this would not cover any other utils or values, such as your last example.

**Reply by gabro** [reply]
I think I'll work under the assumption that all modules `Foo` have a corresponding `foo` export in the index. Seems to hold true in practice, maybe @gcanti can confirm?

**Comment by gabro** [comment]
Done in #29
